### PR TITLE
Add note to describe the format of UUID

### PR DESCRIPTION
**Pull request by paniagua**

### CODE REVIEW

File: apiary.apib [hunk 1]
```diff
@@ -60,6 +60,12 @@ All timestamps are formatted to ISO-8601
 ```shell
 YYYY-MM-DDTHH:MM:SSZ
 ```
+### UUID
+All UUIDs should be in the form of lowercase strings
```
Comment (indrekj) [inline apiary.apib]: you might want to link here https://en.wikipedia.org/wiki/Universally_unique_identifier .

**Comment by indrekj** [comment]
+1

File: apiary.apib (hunk 1 above)
Comment (willmore) [inline apiary.apib]: Do we really have the lowercase constraint?
Also, might want to reference https://www.ietf.org/rfc/rfc4122.txt. Note, that supports both cases.

Comment (paniagua) [inline apiary.apib]: yes, we have the constraint at the moment. I'll reference Indrek's link for now.

Comment (deiwin) [inline apiary.apib]: `should be in the form of` -> `are`

File: apiary.apib [hunk 2]
```diff
@@ -60,6 +60,12 @@ All timestamps are formatted to ISO-8601
...
+
+```shell
```
Comment (deiwin) [inline apiary.apib]: `shell`?

**Comment by paniagua** [comment]
updated

**Comment by deiwin** [comment]
Hmm, I don't see (appropriately, I think) UUIDs currently being mentioned anywhere in the doc. Why is this not necessary? I think we can get away with being more vague and saying that our IDs are strings, as we're doing right now, without having to specify which format we're using.

**Comment by paniagua** [comment]
we had a client trying to use capital case - so not really getting away with being vague.

**Comment by deiwin** [comment]
Well, they should only be using values directly returned by us? If they change the values in any way (e.g. changing case) than they shouldn't expect it to work.

**Comment by paniagua** [comment]
the limitation for the UUIDs in lower case still exist. Even if somebody is just testing things around the clearer we are the better.

**Comment by justindipietro** [comment]
Its better to explain more (remember, clients do not have the best developers a lot the time) so we can prevent as much as possible from ending up in our ticket system.

**Comment by deiwin** [comment]
We don't allow the IDs to be created externally, do we? So the lower-case limitation doesn't really matter, because they shouldn't be defining their own anywhere, anyhow. I don't think there's really a reason to accommodate in this case.

It's just that by being more specific, we're also setting up more restrictions for ourselves.

**Comment by deiwin** [comment]
Maybe there's some other underlying misunderstanding we could remedy. Can I find information about the particular client issue somewhere?

**Comment by paniagua** [comment]
I don't see how saying that UUIDs are lowercase are more restrictions and it just describes the current situation where we are.

**Comment by deiwin** [comment]
Clients might add `.toLowerCase()` to the IDs in all their API request based on this doc, to be safe or whatever. And now suddenly we can't send (actually accept as input, but pretty much the same) anything that isn't strictly lowercase, without braking backwards compatibility.

**Comment by indrekj** [comment]
There might be IDs that aren't UUIDs. You could say: "Our IDs are case sensitive Strings". Although I see Deiwin point. Only we are generating these ID's so how do they even get to be upcase?

**Comment by indrekj** [comment]
btw postgres handles upcase/downcase UUIDs correctly. I'm not sure if we have different checks elsewhere.

**Comment by paniagua** [comment]
I'll update it to case sensitive.

**Comment by deiwin** [comment]
> I'll update it to case sensitive.

I still object.

**Comment by paniagua** [comment]
Problem:

While trying to create an operator the list of sites had to be in lower case otherwise the api think the operator who is creating the new operator does not have access to the site 

@indrekj @deiwin what is the solution then?

**Comment by indrekj** [comment]
`the list of sites had to be in lower case`. Where did you get list of sites with uppercase IDs?

**Comment by paniagua** [comment]
what if the developer upper-cased the ID - just because is not he brightest

**Comment by deiwin** [comment]
> what if the developer upper-cased the ID - just because is not he brightest

This is like saying "what if the developer removed the dashes in the ID". They're doing something stupid, just respond with a 404.

**Comment by indrekj** [comment]
Then he'll get an error response. Lets say github says your issue id is `5`. you don't try to do `GET /issues/five` instead of `GET /issues/5`. ID is identifier, you're not supposed to change it.

So the use case is: Developer fetches all sites. Then he iterates over it and calls `#upcase` on the IDs and then creates a request it. Maybe we're not supposed to support everything :).

**Comment by deiwin** [comment]
@paniagua, can you elaborate what actually went wrong in this particular case? Why did they try to use uppercase IDs?

**Comment by paniagua** [comment]
I don't know why they tried to use the upper case - and ideally i should not know. 

Wouldn't something in the troubleshooting section saying "if you are querying for `/issues/five` you are doing it wrong, you should query for `/issues/5` help. All what i am saying is that I feel that it is in our benefit to help - we can't expect every one to be smart. 

BTW @indrekj  you mentioned that Postgresql manages upper and lower case. Does it mean something is not working correctly? - because at the moment it just worked with lowercase :(

**Comment by indrekj** [comment]
PG is using UUID's but in the services we are still using Strings which means they are case sensitive.

I looked around in few APIs (Github, twilio). I don't see them mentioning this kind of information.

We can't prevent everything. We also have some tokens which must be case sensitive. We can't just say everything is lowercase. Then there are dashes that you can't remove. What if somebody puts a space before or after the id, etc. It should be quite obvious that if you get 404 then you provided an invalid ID.

**Comment by deiwin** [comment]
> I don't know why they tried to use the upper case - and ideally i should not know.

I agree, that we shouldn't know, but I'm trying to understand what went wrong, so we can be sure that we're fixing the correct issue, not just one of perhaps many effects of some underlying issue. Like IDs not being handled programmatically and instead passed on by email or IM or some such and then entered manually by someone somewhere.

> All what i am saying is that I feel that it is in our benefit to help - we can't expect every one to be smart.

Yes, we should help. But we want to avoid setting unnecessary restriction on ourselves. As an example, saying that IDs may not be modified, although sounding redundant, gives the same information, but doesn't set any additional restrictions on us.

**Comment by paniagua** [comment]
> It should be quite obvious that if you get 404 then you provided an invalid ID.

you can hope :) - how frequent common sense is used 

> doesn't set any additional restrictions on us.

seems we have the restrictions anyways 

If you both feel that strong about it then ok - i'll agree to disagree

**Pull request closed without merging**